cli/src/program.rs

```diff
@@ -72,18 +80,28 @@ pub const CLOSE_PROGRAM_WARNING: &str = "WARNING! Closed programs cannot be recr
 pub enum ProgramCliCommand {
     Deploy {
         program_location: Option<String>,
+        fee_payer_signer_index: SignerIndex,
         program_signer_index: Option<SignerIndex>,
-        program_pubkey: Option<Pubkey>,
         buffer_signer_index: Option<SignerIndex>,
-        buffer_pubkey: Option<Pubkey>,
         upgrade_authority_signer_index: SignerIndex,
         is_final: bool,
         max_len: Option<usize>,
         allow_excessive_balance: bool,
         skip_fee_check: bool,
```

Comment:
For context, on `master` for some arguments we need to pass both `signer` and `pubkey` (for the same signer) into `process_program_deploy`. Instead what I've done in this PR is - removed the need to pass those additional `pubkey(s)` and replaced them with `NullSigner(s)` (we still have to pass around a bunch of signers, or rather their indexes - no way around that).

This refactoring allows for:
- (1) removing those unnecessary pubkey args in functions like `process_program_deploy` (which typically, but not always, are duplicates of what we have among signers passed in another arg)
- (2) treating cases when cli argument can be either a signer or a pubkey **same way**, both when parsing cli argument and when using it in functions like `process_program_deploy`, `process_program_upgrade`, ... (for example, `--fee-payer` is treated universally as Signer - regardless of whether it is an `ONLINE_SIGNER` or `ONLINE_SIGNER_PUBKEY`)

(1)st on it's own is questionable (since it kinda makes it a bit more magical and implicit),
(2)nd results in more readable code overall (instead of having duplicate ad-hoc if-check all over the place) - and hence is the reason why I'm proposing this change.

If I were to "revert this change" or even just move it into a separate PR - I'd need to implement an alternative way for `process_program_upgrade` to work with both `ONLINE_SIGNER` or `ONLINE_SIGNER_PUBKEY` (since online signer  pubkey is needed for `--sign-only` mode to work), so
- if you don't like this change ^ overall, I guess I'll just have to throw it away and reimplement it in accordance with the current code (although I personally very much dislike the way `config.signers[0]` is being referenced there all over the place - implying certain assumptions were made)
- if you do like the change but just want to review it separately from this PR, I can make a "preliminary PR with that refactoring only" and simply rebase this one on top of it (to avoid extra work of implementing a separate solution here)

Reply:
Unless I'm missing a subtlety, the refactor isn't necessary since the logic is already there to do what you want (ie. provide pubkeys as signers), but we can have that discussion in the other comment.

As for using `config.signers[0]`, it's definitely not ideal, so I would be happy to review changing that in a separate PR!

Reply:
> As for using config.signers[0], it's definitely not ideal, so I would be happy to review changing that in a separate PR!

I'll move all the logic related to `deploy` command into separate PR(s) - so it doesn't complicate other things, let's discuss `upgrade`-related stuff here,

Reply:
> Unless I'm missing a subtlety, the refactor isn't necessary since the logic is already there to do what you want (ie. provide pubkeys as signers), but we can have that discussion in the other comment.

Let me describe the subtlety I'm trying to address (which `process_transfer` is also susceptible to, and I'm using it in my example below):

take a look at this code:

https://github.com/norwnd/solana/blob/372b86f40f14bf76f987db3751ecf081d5537a5e/cli/src/wallet.rs#L566-L573

and suppose (for whatever reason) the user specified `from` **signer** and also `fee_payer` to be a **pubkey** of that same signer (which'll resolve to `NullSigner`) - so it matches `from_pubkey` value. Maybe user was lazy to specify `fee_payer` as signer, maybe he was just playing around with cli trying to test some stuff. The issue with this is that now on master we don't give non-null signers higher priority over null signers in `generate_unique_signers` ... which means `process_transfer` **might** get a NullSiger for `from` argument and it certainly doesn't expect that (hence it probably fails in some unpredictable manner, unless we are lucky and it's accidentally taken care of).

Now, perhaps for `process_transfer` it's an unlikely case to happen in practice. The thing is, we might hit more "valid" use-case in the future (more similar to `--signer` stuff you mentioned in https://github.com/solana-labs/solana/pull/33860#discussion_r1414393221) and forget about this subtlety by that time. That's why I typically like to solve things like this in "generic way" when it's easy to do so - so that code I write today is future proof.

I'll wait for you reply on this, let me know if you think this is something worth addressing. If it's an overkill in your opinion I'll revert this change and target `--signer` case specifically.

Reply:
In my view, if someone improperly specifies their command-line arguments by using a pubkey in one place and a keypair somewhere else, the tool should loudly say that they did something wrong and that they require a signer everywhere rather than silently promote the pubkey to be a signer.

By promoting pubkeys to signers in certain cases, we'll open ourselves up to people using the tool incorrectly, or worse, using it unsafely. Loud failures are best if the user is doing something strange.

Either way, this is all unrelated to the work in this PR, so we can discuss it more elsewhere, but I find this capability potentially dangerous and unnecessary. Other people might feel differently though!

Reply:
Yeah, strict input validation is another way to address it,

anyway - it isn't really specific to this PR, so I'll remove these changes from it, but first

I've sketched out a PR with the changes related to `deploy` command (in particular, introducing `--fee-payer` and removing direct references to `config.signers[0]` in a couple of places) https://github.com/solana-labs/solana/pull/34341 - lets try to get it merged first, so I can maybe cleanly rebase preserving current commit history in this PR (effectively removing unrelated stuff from this PR).

Reply:
Should be no changes to `deploy`/`write-buffer` left in this PR now.